libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -153,10 +155,135 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
 }
 
-string IRGenerationContext::internalDispatch(size_t _in, size_t _out)
+void IRGenerationContext::setInternalDispatchCandidates(InternalDispatchMap _internalDispatchCandidates)
+{
+	solAssert(internalDispatchClean(), "");
+
+	m_internalDispatchCandidates = move(_internalDispatchCandidates);
+}
+
+tuple<InternalDispatchMap, InternalDispatchMap> IRGenerationContext::consumeInternalDispatchMap()
+{
+	solAssert(
+		m_dispatchableInternalFunctionReferences.empty(),
+		"You must call moveCollectedReferencesToDispatch() before constructing internal dispatch map."
+	);
+
+	InternalDispatchMap internalDispatch = move(m_internalDispatch);
+	InternalDispatchMap internalDispatchCandidates = move(m_internalDispatchCandidates);
+
+	m_internalDispatch.clear();
+	m_internalDispatchCandidates.clear();
+
+	return {move(internalDispatch), move(internalDispatchCandidates)};
+}
+
+void IRGenerationContext::moveCollectedReferencesToDispatch()
+{
+	// First, find (empty) arities newly registered in m_internalDispatch and fill them with
+	// candidates collected so far.
+	for (auto dispatchIt = m_internalDispatch.begin(); dispatchIt != m_internalDispatch.end(); ++dispatchIt)
+	{
+		auto candidateIt = m_internalDispatchCandidates.find(dispatchIt->first);
+
+		solAssert(dispatchIt->second.empty() || candidateIt == m_internalDispatchCandidates.end(), "");
+
+		if (candidateIt != m_internalDispatchCandidates.end())
+		{
+			for (FunctionDefinition const* function: candidateIt->second)
+				enqueueFunctionForCodeGeneration(*function);
+
+			dispatchIt->second = move(candidateIt->second);
+			m_internalDispatchCandidates.erase(candidateIt);
+		}
+	}
+
+	// Now process the references, adding them either as candidates or as dispatch members,
+	// depending on whether the arity has been registered or not.
+	for (auto function: m_dispatchableInternalFunctionReferences | boost::adaptors::map_values)
+	{
+		solAssert(function, "");
+		Arity arity = functionArity(*function);
+
+		auto dispatchIt = m_internalDispatch.find(arity);
+		auto candidateIt = m_internalDispatchCandidates.find(arity);
+		solAssert((dispatchIt == m_internalDispatch.end()) || (candidateIt == m_internalDispatchCandidates.end()), "");
+
+		if (dispatchIt != m_internalDispatch.end())
+		{
+			dispatchIt->second.insert(function);
+			enqueueFunctionForCodeGeneration(*function);
+		}
+		else if (candidateIt != m_internalDispatchCandidates.end())
+			candidateIt->second.insert(function);
+		else
+			m_internalDispatchCandidates[arity] = {function};
+	}
+	m_dispatchableInternalFunctionReferences.clear();
+}
+
+string IRGenerationContext::collectDispatchableReference(Expression const& _expression, FunctionDefinition const& _function)
+{
+	solAssert(m_dispatchableInternalFunctionReferences.count(&_expression) == 0, "");
+
+	m_dispatchableInternalFunctionReferences[&_expression] = &_function;
+	return internalDispatchFunctionName(functionArity(_function));
```

Comment:
I don't see any use of the return value.

Reply:
Right. This is a leftover from before one of the rewrites when I tried to use an earlier version of this function in `endVisit(FunctionCall)` where the name would have been used. I'll remove it.

In general I don't really like that these functions in the context are returning function names anyway. I'd prefer them to be separate calls but I'm just following a convention used in the context.

Reply:
Done.